File: packages/sdk-ts/src/viz/health.ts
/**
 * Health report and impact analysis — standalone functions operating on ManifestInstance.
 *
 * Extracted from ManifestInstance to keep the kernel class focused.
 */

import type { ManifestInstance } from "../kernel/instance.js";
import type { KindPort } from "../kernel/protocols.js";
import type { CompositionProfile } from "../kernel/composition-resolver.js";

// ---------------------------------------------------------------------------
// Helpers (private access via cast)
// ---------------------------------------------------------------------------

function _kinds(mi: ManifestInstance): Map<string, KindPort> {
  return (mi as any)._kinds;
}

function _profiles(mi: ManifestInstance): readonly CompositionProfile[] {
  return (mi as any)._profiles;
}

// ---------------------------------------------------------------------------
// Health Report
// ---------------------------------------------------------------------------

export function healthReport(mi: ManifestInstance): Record<string, unknown> {
  const kinds = _kinds(mi);
  const profiles = _profiles(mi);

  const tree = mi.dependencyTree();
  const comp = mi.compositionResult;

  // All orchestrators (agents) — found via composition profiles
  const allAgents = mi.allWhere(kp => kp.isPromptTarget && !kp.isRoot && !kp.flattenInContext);
  const agentNames = new Set(allAgents.map((a) => a.name));

  // --- Generic slot-based health checks via CompositionProfile ---
  const slotIssues: Record<string, string[]> = {};
  const slotCoverage: Record<string, string> = {};
  const slotTotals: Record<string, number> = {};
  let hasIssues = false;

  for (const agent of allAgents) {
    const profile = mi.profileFor(agent);
    if (!profile) continue;

    for (const slot of profile.slots) {
      if (!slot.healthCheck) continue;
      const deps = mi.iterDocDeps(agent);
      const slotDeps = deps.find(d => d.label === slot.name);
      const names = slotDeps?.names ?? [];

      if (slot.healthCheck.rule === "at-least-one" && names.length === 0) {
        const key = slot.healthCheck.issueKey;
        (slotIssues[key] ??= []).push(agent.name);
        hasIssues = true;
      }
    }
  }

  // Coverage stats per slot
  for (const profile of profiles) {
    for (const slot of profile.slots) {
      if (!slot.healthCheck) continue;
      const issueList = slotIssues[slot.healthCheck.issueKey] ?? [];
      const covered = allAgents.length - issueList.length;
      slotCoverage[slot.name] = `${covered}/${allAgents.length} agents`;

      // Count total docs of this slot's target kind
      const targetKp = mi.kindForAlias(slot.targetAlias);
      if (targetKp) {
        const allOfKind = mi._all(targetKp.kind);
        slotTotals[`${slot.name}_total`] = allOfKind.length;

        // For kinds with graphMeta returning severity, count rules
        let rulesCount = 0;
        for (const d of allOfKind) {
          const meta = targetKp.graphMeta?.(d) ?? targetKp.summary(d) ?? {};
          const m = meta as Record<string, unknown>;
          if (typeof m.rules === "number") rulesCount += m.rules;
        }
        if (rulesCount > 0) slotTotals[`${slot.name}_rules_total`] = rulesCount;

        // has-error-severity: find agents missing error-severity docs
        if (slot.healthCheck.rule === "at-least-one") {
          const errorDocs = allOfKind.filter(d => {
            const m = (targetKp.graphMeta?.(d) ?? targetKp.summary(d) ?? {}) as Record<string, unknown>;
            return m.severity === "error";
          });
          if (errorDocs.length > 0) {
            const agentsWithError = new Set<string>();
            for (const agent of allAgents) {
              const deps = mi.iterDocDeps(agent);
              const slotDeps = deps.find(dd => dd.label === slot.name);
              for (const depName of slotDeps?.names ?? []) {
                if (errorDocs.some(ed => ed.name === depName)) {
                  agentsWithError.add(agent.name);
                }
              }
            }
            const missing = [...agentNames].filter(n => !agentsWithError.has(n)).sort();
            if (missing.length > 0) {
              slotIssues[`agents_missing_error_${slot.name}`] = missing;
            }
          }
        }
      }
    }
  }

  // Referenced deps (from tree)
  const referenced = new Set<string>();
  for (const info of Object.values(tree)) {
    const treeInfo = info as Record<string, unknown>;
    const dependsOn = treeInfo.depends_on as Record<string, Record<string, unknown>> | undefined;
    if (!dependsOn) continue;
    for (const deps of Object.values(dependsOn)) {
      for (const depName of Object.keys(deps)) {
        referenced.add(depName);
      }
    }
  }

  // Find orphan documents (not referenced, not root, not prompt-target)
  const orphans: Array<{ kind: string; name: string }> = [];
  for (const doc of mi.documents) {
    const _kp = kinds.get(`${doc.apiVersion}\0${doc.kind}`);
    if (_kp?.isRoot || _kp?.isPromptTarget) continue;
    if (!referenced.has(doc.name)) {
      orphans.push({ kind: doc.kind, name: doc.name });
    }
  }

  const isHealthy = comp.missing.length === 0 && !hasIssues;

  return {
    status: isHealthy ? "healthy" : "warnings",
    composition_valid: comp.missing.length === 0,
    missing_refs: comp.missing,
    agents_total: allAgents.length,
    // Slot issues — each key comes from slot.healthCheck.issueKey
    ...slotIssues,
    // Slot totals (e.g. guardrails_total, guardrails_rules_total)
    ...slotTotals,
    orphan_documents: orphans,
    coverage: slotCoverage,
  };
}

// ---------------------------------------------------------------------------
// Impact Analysis
// ---------------------------------------------------------------------------

export function impact(mi: ManifestInstance, kind: string, name: string): Record<string, unknown> {
  const tree = mi.dependencyTree();
  const affected: Array<{ agent: string; relationship: string }> = [];

  for (const [agentName, info] of Object.entries(tree)) {
    const treeInfo = info as Record<string, unknown>;
    const dependsOn = treeInfo.depends_on as Record<string, Record<string, unknown>> | undefined;
    if (!dependsOn) continue;
    for (const [depType, deps] of Object.entries(dependsOn)) {
      if (name in deps) {
        affected.push({ agent: agentName, relationship: depType });
      }
    }
  }

  const doc = mi._one(kind, name);
  return {
    kind,
    name,
    description: doc ? ((doc.metadata as Record<string, unknown>).description ?? "") : "",
    found: doc !== null,
    affected_agents: affected,
    affected_count: affected.length,
    safe_to_remove: affected.length === 0,
  };
}

File: packages/sdk-py/dna/viz/health.py
"""Health report and impact analysis — standalone functions operating on ManifestInstance.

Extracted from ManifestInstance to keep the kernel class focused.
1:1 parity with TypeScript viz/health.ts.
"""
from __future__ import annotations
from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    from dna.kernel.instance import ManifestInstance


def health_report(mi: ManifestInstance) -> dict[str, Any]:
    """Health report: coverage gaps, orphan documents, missing refs.

    Iterates CompositionProfile slots generically — no kind names
    hardcoded. Each slot with a health_check hint generates issue
    lists automatically.
    """
    tree = mi.dependency_tree()
    comp = mi.composition_result

    # All orchestrators (agents)
    all_agents = mi.all_where(
        lambda kp: kp.is_prompt_target
        and not getattr(kp, "is_root", False)
        and not getattr(kp, "flatten_in_context", False)
    )
    agent_names = {a.name for a in all_agents}

    # --- Generic slot-based health checks via CompositionProfile ---
    slot_issues: dict[str, list[str]] = {}
    slot_coverage: dict[str, str] = {}
    slot_totals: dict[str, int] = {}
    has_issues = False

    for agent in all_agents:
        profile = mi.profile_for(agent)
        if not profile:
            continue
        for slot in profile.slots:
            if not slot.health_check:
                continue
            deps = mi.iter_doc_deps(agent)
            slot_deps = next((d for d in deps if d["label"] == slot.name), None)
            names = slot_deps["names"] if slot_deps else []

            if slot.health_check.rule == "at-least-one" and not names:
                key = slot.health_check.issue_key
                slot_issues.setdefault(key, []).append(agent.name)
                has_issues = True

    # Coverage stats + totals per slot
    for profile in mi._profiles:
        for slot in profile.slots:
            if not slot.health_check:
                continue
            issue_list = slot_issues.get(slot.health_check.issue_key, [])
            covered = len(all_agents) - len(issue_list)
            slot_coverage[slot.name] = f"{covered}/{len(all_agents)} agents"

            target_kp = mi.kind_for_alias(slot.target_alias)
            if target_kp:
                all_of_kind = mi._kernel.query_list_sync(mi.scope, target_kp.kind)
                slot_totals[f"{slot.name}_total"] = len(all_of_kind)

                # Count rules via graphMeta/summary
                rules_count = 0
                for d in all_of_kind:
                    meta = {}
                    gm = getattr(target_kp, "graph_meta", None)
                    if callable(gm):
                        meta = gm(d) or {}
                    elif hasattr(target_kp, "summary") and callable(target_kp.summary):
                        meta = target_kp.summary(d) or {}
                    r = meta.get("rules")
                    if isinstance(r, int):
                        rules_count += r
                if rules_count > 0:
                    slot_totals[f"{slot.name}_rules_total"] = rules_count

                # Error-severity coverage check
                error_docs = []
                for d in all_of_kind:
                    meta = {}
                    gm = getattr(target_kp, "graph_meta", None)
                    if callable(gm):
                        meta = gm(d) or {}
                    elif hasattr(target_kp, "summary") and callable(target_kp.summary):
                        meta = target_kp.summary(d) or {}
                    if meta.get("severity") == "error":
                        error_docs.append(d)

                if error_docs:
                    agents_with_error: set[str] = set()
                    for agent in all_agents:
                        deps = mi.iter_doc_deps(agent)
                        slot_deps = next((dd for dd in deps if dd["label"] == slot.name), None)
                        for dep_name in (slot_deps["names"] if slot_deps else []):
                            if any(ed.name == dep_name for ed in error_docs):
                                agents_with_error.add(agent.name)
                    missing = sorted(agent_names - agents_with_error)
                    if missing:
                        slot_issues[f"agents_missing_error_{slot.name}"] = missing

    # Referenced deps (from tree)
    referenced: set[str] = set()
    for info in tree.values():
        for deps in info.get("depends_on", {}).values():
            referenced.update(deps.keys())

    # Find orphan documents
    orphans: list[dict[str, str]] = []
    for doc in mi.documents:
        kp = mi._kinds.get((doc.api_version, doc.kind))
        if mi._is_root_doc(doc) or getattr(kp, "is_prompt_target", False):
            continue
        if doc.name not in referenced:
            orphans.append({"kind": doc.kind, "name": doc.name})

    is_healthy = comp.valid and not has_issues

    return {
        "status": "healthy" if is_healthy else "warnings",
        "composition_valid": comp.valid,
        "missing_refs": comp.missing,
        "agents_total": len(all_agents),
        **slot_issues,
        **slot_totals,
        "orphan_documents": orphans,
        "coverage": slot_coverage,
    }


def impact(mi: ManifestInstance, kind: str, name: str) -> dict[str, Any]:
    """Analyze the impact of changing or removing a document.

    Returns which agents depend on this document and how.
    """
    tree = mi.dependency_tree()
    affected: list[dict[str, Any]] = []

    for agent_name, info in tree.items():
        for dep_type, deps in info.get("depends_on", {}).items():
            if name in deps:
                affected.append({
                    "agent": agent_name,
                    "relationship": dep_type,
                })

    # Check if other non-agent docs reference it (reverse lookup)
    doc = mi._kernel.get_document_sync(mi.scope, kind, name)
    return {
        "kind": kind,
        "name": name,
        "description": doc.metadata.get("description", "") if doc else "",
        "found": doc is not None,
        "affected_agents": affected,
        "affected_count": len(affected),
        "safe_to_remove": len(affected) == 0,
    }
